enhancements/etcd/dr-leveraging-ceo-scaling.md

```diff
@@ -0,0 +1,159 @@
+1. Get the first etcd member running. 
+2. Get the operator to deploy static pods that can decide that they are being restored, so they dont start on the existing data directories.
+
+### Static Pod identifying a restore
+
+Assuming the first member is up, the control plane will be up and operator start running. A user can trigger a new rollout of the etcd pods by running `oc patch forceredpeloymment=reason-time`. New pods will be rolled out on the remaining masters that have to be restored. Before etcd commands on these pods are run, the existing data directory needs to be removed, because they contain the previous cluster state information.
+
+One possible solution is, during the start up pod queries etcd membership, if the pod is an unstarted member, it can be assumed that the data directory present is from previous cluster and moved out.
+
+Other solution is to check the cluster id of the new cluster with the cluster id of the existing data directory, move out the data directory only if they are different. This is a little time consuming to implement and test.
+
+### Getting the first etcd member running
+
+There are two alternative for getting the first etcd member running:
+
+ 1. Admin uses a command to render a bootstrap like pod with recovered data directory. The workflow here would be to:
+    1. move the existing etcd pod yaml out of kubelet manifest directory and move etcd data directory into a different location.
+    2. get the etcd ETCD_NAME and ETCD_PEER_URL values from backed up etcd pod yaml 
+    3. use etcdctl to restore the snapshot 
+    4. start the etcd cert signer
+    5. reuse the CEO binary render command to generate a pod yaml. 
+    6. start the pod on a port different from regular etcd server, peer and metrics ports mount it the new restored data directory
+    7. update the KAS config to point to new etcd server and restart the kube-apiserver pod
+    8. create a new host-etcd endpoint resource to point to the new etcd server
+    9. force CEO redeployment via `oc patch forceredpeloymment=reason-time`
+    10. wait for scale up to complete
+    11. clean up: stop etcd-cert signer, remove the temporary etcd pod and its data directory
+    
+ Challenges involved: making the render command work, making sure new ports can be opened up on masters to bind the temporary etcd and cert-signer.
+
+ 2. Use the backed up etcd-pod yaml and static pod resources to start the etcd pod. When the backup scripts runs, it includes the latest static pod resources (yaml, secrets and configmaps). The workflow here would be:
+    1. move the existing etcd pod yaml out of kubelet manifest directory and move etcd data directory into a different location.
+    2. get the etcd ETCD_NAME and ETCD_PEER_URL values from etcd pod yaml
+    3. use etcdctl to restore the snapshot 
+    4. modify the backed up pod yaml to remove all the bash and replace two env variables from step 1.
```

Comment:
this looks pretty clean.  Could we actually produce the pod.yaml we need and lay it down dormant in a configmap?  Making the action on the node a straight copy/paste (via a script perhaps).

Reply:
You can imagine making a configmap with nodeA-backup-pod, nodeB-backup-pod, nodeC-backup-pod.

Reply:
+1, I can imagine making a pod YAML and laying it out beforehand. Infact, since the env variable created in step 2 will be consumed by etcd in the pod, we can make the pod yaml to look like:

1. check if data-directory exists, if it does start etcd.
2. use etcdctl to restore the snapshot
3. start the etcd with the right env variables for single-member cluster.

With this, the admin steps are modified as:
1.  move the existing pod yaml out of the kubelet manifest directory
2. move the existing data directory into a backup location or delete it
3. move the restore pod yaml laid out by the configmap into kubelet manifest directory
4. restart kubelet
5. force CEO redeployment by `oc patch forceredpeloymment=reason-time`
6. When CEO redeployment is forced, the existing nodes will start with new pods, a scale-up similar to bootstrap will take place
8. CEO will redeploy the modified pod on the recovery master node to the current revision. The pod yaml laid by CEO pod should have the logic to move the recovery pod laid out by configmap out of the manifest directory if it exists and start itself